lib/proxy/log_buffer/decorator.rb

```diff
@@ -1,32 +1,53 @@
 require 'proxy/log_buffer/decorator'
 require 'proxy/log_buffer/buffer'
+require 'thread'
 
 module Proxy::LogBuffer
   class Decorator
+    def self.instance
+      @@instance ||= new(::Proxy::LoggerFactory.logger)
+    end
+
     def initialize(logger, buffer = Proxy::LogBuffer::Buffer.instance)
       @logger = logger
       @buffer = buffer
+      @mutex = Mutex.new
+      @log_rolled = false
+    end
+
+    # due to synchronization can't re-open the log from the signal trap
+    def log_rolled
+      @log_rolled = true
+    end
+
+    def handle_log_rolling
+      need_rolling = !@logger.instance_of?(::Syslog::Logger) rescue true
+      @logger = ::Proxy::LoggerFactory.logger if need_rolling
+      @log_rolled = false
     end
 
     def add(severity, message = nil, progname = nil, backtrace = nil)
-      severity ||= UNKNOWN
-      if message.nil?
-        if block_given?
-          message = yield
-        else
-          message = progname
+      @mutex.synchronize do
```

Comment:
That, but mainly because @logger is referred to more than once. The mutex guarantees us that all writes are done to the *same* logger. For example, without this synchronization it would be possible for https://github.com/theforeman/smart-proxy/pull/394/files#diff-6ae09024c8b8cebdf00278e6e789bc93R41 and https://github.com/theforeman/smart-proxy/pull/394/files#diff-6ae09024c8b8cebdf00278e6e789bc93R42 to end up in different log files. There's also the awkwardness of multiple re-initialization of @logger in the event of log rolling.

It's not completely thread-safe atm (I don't think boolean assignment and checks are atomic), but the mutex makes the whole thing more deterministic (which in turns hurts my head less).